### PR TITLE
Support mock generation for generic interfaces

### PR DESCRIPTION
**Pull request by jonathansharman**
### Dependencies

None.

### Documentation

- [🐛 mock can't handle generic interfaces](https://app.asana.com/0/1201010009456007/1206975799710921/f)
- Discovered in https://github.com/nicheinc/pgsql/pull/19#discussion_r1531006944

### Description

This adds support for generating mocks for generic interfaces - i.e. interfaces with type parameters. Currently, attempting to do so produces a mock struct with no type parameters, which (if the types are referenced in any of the methods) tends to produce compilation errors.

Generated mock files currently include assignment of the mock type to a variable of the interface type, as a kind of unit test that the mock actually implements the interface. It's unclear how in general we should generate these checks in the case of generic mocks. @bclarkx2 included implementation checks in the [mocks that he fixed by hand](https://github.com/nicheinc/pgsql/blob/a3e869832cd72cab196c78ec1b3fa501bf124ba2/exp/pgrows/async_mock.go#L20) in his `pgsql` branch, using `any` as the type arguments. That's a very reasonable choice in that case since the type parameters are only constrained to `any`. However, if there are nontrivial constraints, it's much less clear what types we should use. Rather than attempt to produce an arbitrary implementing type for each parameter, I have simply omitted these checks for generic mocks. That shouldn't have any real effect on users.

This PR also upgrades dependencies.

### Testing Considerations

1. In `pgsql@x/scan-rows`, (temporarily) change the branch names in the `go:generate` directives from `mock@main` to `mock@fix/generic-interfaces`
2. Regenerate the mocks with `go generate ./...`
3. Confirm that the mock structs have correct type parameters

### Versioning

Patch (but `mock` is currently unversioned).

### CODE REVIEW

File: example/myInterface.go [hunk 1]
```diff
@@ -22,6 +25,7 @@ type MyInterface interface {
 	NamedParam(str string)
 	NamedVariadicParam(strs ...string)
 	SameTypeNamedParams(str1, str2 string)
+	InternalTypeParam(internal internal.Internal)
```
Comment (jonathansharman) [inline example/myInterface.go:28]: While I was here, I added another method to check how `mock` handles local/internal types.

File: iface/get.go [hunk 2]
```diff
@@ -71,6 +71,19 @@ func GetInterface(dir, ifaceName string) (Interface, error) {
 		Package: pkg.Name,
 		Name:    ifaceObj.Name(),
 	}
+	qualifier := Qualify(pkg.Types, imps, &iface.Imports)
+
+	// Record type parameter list info.
+	if ifaceNamed, ok := ifaceObj.Type().(*types.Named); ok {
+		typeParams := ifaceNamed.TypeParams()
```
Comment (jonathansharman) [inline iface/get.go:78]: There is a `validateType` function in this file to recursively validate all the types involved in the interface. This function includes the case where the [type is `*types.Named`](https://github.com/nicheinc/mock/blob/a68fb3801a78dfbd7e0c4d8cc294779929fd84ff/iface/get.go#L297-L302), and I considered validating the type constraints at that point. However, the error log is currently commented out, and the default case, which in theory is the only way this function can return "not valid", instead returns "valid".

So in short, this function doesn't do anything - it will always just return `true`, for any type. I don't understand the code well enough to be comfortable removing it entirely, and changing the `default` case to return "not valid" causes the well-formed example types to fail. 😞 So I'm just going to leave it alone for now.

Comment (nathanjcochran) [reply iface/get.go:78]: > the default case, which in theory is the only way this function can return "not valid", instead returns "valid".

It's been awhile since I've looked at the code, but I think [this](https://github.com/nicheinc/mock/blob/a68fb3801a78dfbd7e0c4d8cc294779929fd84ff/iface/get.go#L251) is actually the main way the function can return "not valid".

Comment (jonathansharman) [reply iface/get.go:78]: Wow, I don't know how I glossed over that case. :see_no_evil: Maybe I'll try to incorporate those constraint types after all then. Also, hey Cochran! :wave:

Comment (nathanjcochran) [reply iface/get.go:78]: Hey!! 👋

Looking forward to the PRs for merging these changes upstream 😉😉😉

Comment (nathanjcochran) [reply iface/get.go:78]: And just for clarity, the purpose of `ValidateType` is to check whether any of the type errors in the package are actually related to the interface you're trying to mock. It makes it possible to generate mocks for an interface even if there are other unrelated errors in the package, and only prevents mock generation if any of the errors are actually relevant to the interface definition.

Originally, the tool would error out and fail to generate a mock if `len(pkg.Errors) > 0`, but that meant you couldn't generate a mock if you had _any_ errors in the package. I often wanted to (re-)generate a mock when I was in the middle of work (and other stuff was broken), so I changed it to stop erroring out when `len(pkg.Errors) > 0` and instead added `ValidateType`, so that  I could determine if any of the types actually involved in the interface were broken/invalid, and only fail to generate a mock in that case (but it still [returns all of the package errors](https://github.com/nicheinc/mock/blob/a68fb3801a78dfbd7e0c4d8cc294779929fd84ff/iface/get.go#L63), when that happens).

`ValidateType` just recursively descends from more "complex" types to the "basic" types that constitute them (for example, from `*chan[int]` to `chan[int]` to `int`). When it gets to the bottom level, it checks whether the type is `types.Invalid`, which happens if the type isn't defined. Like, I believe you'd hit that case for `chan[Invalid]` if `Invalid` wasn't defined.

Comment (jonathansharman) [reply iface/get.go:78]: Awesome, thanks for that explanation! I think we'd definitely want to include the type parameter constraints in this check then since they need to be repeated in the mocked struct definition, and we wouldn't want to generate a type with broken constraints. :+1:

Comment (nathanjcochran) [reply iface/get.go:78]: Agreed!

**Review by bclarkx2 — CHANGES_REQUESTED** [review]
Awesome!! This looks like a great improvement, can't wait to start using this :rocket: 

I spotted one functional issue I think might be a blocker, and then left one musing about the go version in this new post-go1.21 world.

File: template.go [hunk 3]
```diff
@@ -10,21 +10,23 @@ import (
...
 	T *testing.T
 	{{- range .Methods }}
 	{{ .Name }}Stub func({{ .Params }}) {{ .Results }}
 	{{ .Name }}Called int32
 	{{- end }}
 }
 
+{{ if not .TypeParams }}
 var _ {{ .Name }} = &{{ .Name }}Mock{}
+{{ end }}
```
Comment (bclarkx2) [inline template.go:23]: > Generated mock files currently include assignment of the mock type to a variable of the interface type, as a kind of unit test that the mock actually implements the interface. It's unclear how in general we should generate these checks in the case of generic mocks.

Major caveat: this might be a terrible idea :joy:

So, the idea behind this test we add to the mock files is that we want the compiler to verify that a value of type `MyMock` can be assigned to a variable of type `MyInterface`, thereby proving that the type `MyMock` implements the `MyInterface` interface.

Excuse the abuse of notation, but this is essentially a propositional logic statement:
`MyMock satisfies MyInterface` (If you have a `MyMock`, then you have a `MyInterface`).

When we add a type constraint set `T` to the definition of `MyInterface`, I think the problem is that the desired statement becomes a first order logic expression:
`∀X (X satisfies T) -> MyMock[X] satisfies MyInterface[X]` (For all types X, if X satisfies the type constraints T, then a MyMock[X] will satisfy the concrete interface type MyInterface[X])

I totally agree that deciding what X to pick to create this test isn't super helpful. (For one, actually picking one is hard, and for two, it's hard to say that the selected X would guarantee that the mock would satisfy the interface for all allowed values of X).

However... could we just generate code here that essentially implements the full first order logic expression? :thinking: 

e.g., imagine that given an interface definition `MyInterface[T any, U comparable]`, we generated a snippet like the following:
```go
func _[T any, U comparable]() {
    var _ MyInterface[T, U] = &MyInterfaceMock[T, U]{}
}
```

My feeling is that this wouldn't compile if the mock didn't have the same type constraints as the `MyInterface` type :thinking: I think this would be easy enough to generate, too, since we have `TypeParams.String` and `TypeParams.Names`.

What do you think of this approach?

File: iface/get.go [hunk 4]
```diff
@@ -282,6 +295,10 @@ func validateType(typ types.Type, visited *[]types.Type) bool {
 		return validateType(t.Elem(), visited)
 
 	case *types.Named:
+		typeParams := t.TypeParams()
+		for i := range typeParams.Len() {
+			validateType(typeParams.At(i).Constraint(), visited)
```
[the remark this answers is not in the document]
Comment (jonathansharman) [reply iface/get.go]: My reading comprehension for this function has not been so stellar. :sweat: Indeed, I should be checking that value and short-circuiting. I realized there were other problems too though, which I've (hopefully) fixed in 22db612da1cb5d2d7d95e45cc0be376dc03482d5. (I also switched the for-loops to use Go 1.22 integer-range-for-loops, for fun.)

The previous version didn't drill down into constraint sets to validate each individual type, which means a bogus type within a constraint type set would not be detected. To fix that, I added a [new case to `validateType` for `*types.Union`](https://github.com/nicheinc/mock/blob/22db612da1cb5d2d7d95e45cc0be376dc03482d5/iface/get.go#L295-L301) (which represents a type set) and added code in the `*types.Interface` case to [range over the type's embedded types](https://github.com/nicheinc/mock/blob/22db612da1cb5d2d7d95e45cc0be376dc03482d5/iface/get.go#L283-L287), which is how constraint sets are represented.

I wouldn't be surprised if there are other cases this function doesn't handle yet. :thinking: I wish there were an off-the-shelf way to check "does this type contain any invalid types".

File: go.mod [hunk 5]
```diff
@@ -1,10 +1,10 @@
 module github.com/nicheinc/mock
 
-go 1.21
+go 1.22
```
[the remark this answers is not in the document]
Comment (jonathansharman) [reply go.mod]: Seems good! 1797071c6e61222a94db7d394f974354a4512301 TIL about that `go get go@...` command. :exploding_head:

File: template.go (hunk 3 above)
Comment (jonathansharman) [reply template.go]: Great idea! 2e78e2bb06f6e584ed97dd37c2d1d4f4e8f2fbc3 I also added a generated comment to explain the purpose of these `_` assignments - let me know what you think.

File: iface/get.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (bclarkx2) [reply iface/get.go]: Oh nice, good catch!! I did a quick scan of the `types` package to see if there are any other valid types we could expect to encounter and didn't see any that jumped out... but I agree that it's hard to feel confident that there are no further edge cases :thinking: I suppose the failure mode isn't too bad though -- if an invalid type 'gets through' `mock` will just go ahead and try to make the mock. _That_ procedure may fail with a less helpful error message, but it seems like something we should be able to deal with :thinking:

File: template.go (hunk 3 above)
Comment (bclarkx2) [reply template.go]: Looks great to me, thanks for throwing that in!

**Review by bclarkx2 — APPROVED** [review]
LGTM!! What a generally great, generous generic upgrade :rocket:
